[ENH] Add warning message when overwriting existing role.

daehyun99 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2415 

### List of changes to the codebase in this pull request
- Make Frame for `_GraphRolesMixin`'s test code.
- Write test code for `_GraphRolesMixin` related overwriting existing role.
- Write code logic for `_GraphRolesMixin` related overwriting existing role.

---
- example warning message image.
<img width="1206" height="391" alt="image" src="https://github.com/user-attachments/assets/1a0e00ab-a226-43bd-873c-efa62b0488e9" />

daehyun99 committed:
Test: Write test code for `_GraphRolesMixin` class (#2415)

daehyun99 committed:
Test: add test code for overwriting existing role. && Feat: add warning message when overwriting existing role. (#2415)

daehyun99 committed:
Merge branch 'dev' into 2415/Bug-fix

daehyun99 commented:
@ankurankan I think ready for review.

ankurankan commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env python3
+import warnings
```
We should use the pgmpy logger from `pgmpy.global_vars import logger`. You can find examples of this in the codebase.

ankurankan commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -96,12 +97,24 @@ def with_role(self, role: str, variables, inplace=False):
+                    existing_role = new_graph.nodes(data=True)[var].get("role", None)
+                    if existing_role is not None:
+                        warnings.warn(
+                            f"⚠️ Overwriting existing role for '{var}'. Replacing '{existing_role}' with '{role}'.",
```
I would suggest avoiding the use of special characters to prevent compatibility issues.

ankurankan commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -96,12 +97,24 @@ def with_role(self, role: str, variables, inplace=False):
                 else:
+                    existing_role = new_graph.nodes(data=True)[var].get("role", None)
+                    if existing_role is not None:
+                        warnings.warn(
```
Same here.

ankurankan commented on pgmpy/tests/test_base/test_mixin_roles.py:
```diff
@@ -0,0 +1,117 @@
+
+    def test_outcomes_setter(self):
+        """Test the `outcomes` property setter."""
+        ...
```
These methods seem to be empty. Is this expected?

daehyun99 replied:
Yes.
The test class for `_GraphRolesMixin` was missing, so I created a new testing class, but I feel I don't have enough time lately, so I left it empty.
I expect that by leaving it empty and creating an issue, other contributors will be able to write it.

And i have apply the code review.

Thanks.

---
sample warning message image:
<img width="705" height="522" alt="image" src="https://github.com/user-attachments/assets/76e80e4f-60e7-40d8-b389-cbde046d830f" />

ankurankan reviewed (changes requested):
@daehyun99 Thanks for the PR. I have added a few comments.

daehyun99 committed:
Feat: Apply review #2417. (#2415)

- Change from `warnings` to `pgmpy.global_vars.logger`.
- Avoiding the use of special characters to prevent compatibility issues.

daehyun99 committed:
Merge branch 'dev' into 2415/Bug-fix

ankurankan commented:
@daehyun99 I would suggest removing the test file for now. I think the graph roles methods are being tested in the individual model classes. We can go for a more explicit test in #2427. PR looks good otherwise.

daehyun99 committed:
Test: Delete the file `test_mixin_roles.py` (#2417)

daehyun99 committed:
Merge branch 'dev' into 2415/Bug-fix

daehyun99 commented:
@ankurankan 
Yes, I understand.
I remove the test file.

Thanks.

ankurankan reviewed (approved):
@daehyun99 Looks good. Thanks for the fix.
